### PR TITLE
Enable Unpacking, repacking, signing, and notarization of .pkg files and .app bundles

### PR DESCRIPTION
**Pull request by ellahathaway**
Related to https://github.com/dotnet/arcade/issues/14438

This is the second part of https://github.com/dotnet/arcade/issues/14438. This PR integrates the pkg tool, introduced in https://github.com/dotnet/arcade/pull/15205, into SignTool. More specifically, this PR enables the unpacking and repacking of .pkgs and nested .app bundles within SignTool & adds tests & test data for this functionality into SignTool.Tests.

This is a rather large change, including quite a lot of refactoring as SignTool is becoming a bit bloated. Highlights
- Removed the use of ITaskItem in the signtool sources, aside from the task. It made no sense to have some data passed around as MSBuild items, and others passed around formalized types. Introduce new types where needed.
- Notarization is supported via a "dual sign" mechanism similar to a dual sign certificate. At some point, we may want to change this model and make notarization a separate concept (a new signing information type). For now, I deemed that approach too invasive. Right now, there are special certs for Mac that represent both notarization and signing. When those certs are used. Signtool will both sign and notarize the file, in separate passes.
- Rework the pkg tooling to use the command line API
- Minimally tweak the build manifest model to support the new signing data. I think this should be removed altogether in https://github.com/dotnet/arcade/pull/15415
- Add various tests for new pkg tooling, corner case tests for old bugs.
- If files did not have a matching extension or strong name sign info, they wouldn't get signed even if they had an explicit cert specified. Fix this.

There is more follow on work to be done here. Signtool is due for a major refactor to deal with all the various archive types it now supports. There is far too much "If <archive type>".

### CODE REVIEW

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.proj [hunk 1]
```diff
@@ -58,6 +58,9 @@
 
       <!-- SN is only available on Windows -->
       <SNBinaryPath Condition="$([MSBuild]::IsOSPlatform('Windows'))">$(NuGetPackageRoot)sn\$(SNVersion)\sn.exe</SNBinaryPath>
+      
+      <!-- .pkgs and .app bundle tooling is only available on MacOS -->
+      <PkgToolPath Condition="$([MSBuild]::IsOSPlatform('OSX'))">$(NuGetPackageRoot)microsoft.dotnet.macospkg\$(MicrosoftDotNetSignToolVersion)\tools\net\any\Microsoft.Dotnet.MacOsPkg.dll</PkgToolPath>
```
Comment (mmitche) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.proj:63]:
```suggestion
      <PkgToolPath Condition="$([MSBuild]::IsOSPlatform('OSX'))">$(NuGetPackageRoot)microsoft.dotnet.macospkg\$(MicrosoftDotNetMacOsPkgVersion)\tools\net\any\Microsoft.Dotnet.MacOsPkg.dll</PkgToolPath>
```

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.props [hunk 2]
```diff
@@ -59,6 +59,10 @@
     <FileExtensionSignInfo Include=".zip" CertificateName="None" />
     <FileExtensionSignInfo Include=".tgz" CertificateName="None" />
     <FileExtensionSignInfo Include=".tar.gz" CertificateName="None" />
+    <FileExtensionSignInfo Condition="$([MSBuild]::IsOSPlatform('OSX'))" Include=".pkg" CertificateName="None" />
+    <!-- .app bundles are technically directories, but the Microsoft.DotNet.MacOsPkg
+    tool packs these bundles into zips when unpacking .pkgs -->
+    <FileExtensionSignInfo Condition="$([MSBuild]::IsOSPlatform('OSX'))" Include=".app" CertificateName="None" />
```
Comment (mmitche) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.props:65]: I would leave out the platform checks. This leaves the possibility someone assumes MacOS .pkgs will get signed properly  on Windows, but instead they get ignored. Instead, the pack/unpack process should error out (which it will, since the tool won't work). Then we rely on teams not specifying that .pkgs should be signed on non-Mac platforms.

File: src/Microsoft.DotNet.SignTool.Tests/SignToolTests.cs [hunk 3]
```diff
@@ -1057,6 +1072,141 @@ public void SignZipFile()
...
+            // When .apps are unpacked from .pkgs, they get zipped so they can be signed
+            ValidateFileSignInfos(itemsToSign, strongNameSignInfo, fileSignInfo, s_fileExtensionSignInfo, new[]
+            {
+                "File 'libexample.dylib' Certificate='DylibCertificate'",
+                "File 'test.app' Certificate=''",
+                "File 'WithApp.pkg' Certificate=''",
+            });
+
+            // OSX files need to be zipped first before being signed
+            // This is why the .pkgs and .apps are listed as .zip files below
```
Comment (mmitche) [inline src/Microsoft.DotNet.SignTool.Tests/SignToolTests.cs:1199]: Comment currently out of date?

File: src/Microsoft.DotNet.SignTool/src/FileSignInfo.cs [hunk 4]
```diff
@@ -36,6 +37,18 @@ internal static bool IsMPack(string path)
         internal static bool IsNupkg(string path)
             => Path.GetExtension(path).Equals(".nupkg", StringComparison.OrdinalIgnoreCase);
 
+        // Only recognize and process OSX packages on OSX because
+        // unpacking, repacking, and notarization can only happen on a Mac.
+        internal static bool IsPkg(string path)
+            => Path.GetExtension(path).Equals(".pkg", StringComparison.OrdinalIgnoreCase)
+                && RuntimeInformation.IsOSPlatform(OSPlatform.OSX);
```
Comment (mmitche) [inline src/Microsoft.DotNet.SignTool/src/FileSignInfo.cs:44]: I would leave out these platform checks, and instead let the pack/unpack fail on non-Mac

File: src/Microsoft.DotNet.SignTool.Tests/SignToolTests.cs (hunk 3 above)
Comment (ellahathaway) [reply src/Microsoft.DotNet.SignTool.Tests/SignToolTests.cs:1199]: Looks like it, yes. This comment applies to the changes in https://github.com/ellahathaway/arcade/commit/cd090b8141774c2edbeb7be1d95e00a605cd7444

**Comment by mmitche** [comment]
Converting back to a draft and merging in the latest changes for OSX and strong name signing.

**Comment by mmitche** [comment]
Readddy for review!

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.proj [hunk 5]
```diff
@@ -20,6 +20,7 @@
   </PropertyGroup>
 
   <Import Project="BuildStep.props" />
+  <Import Project="TargetFrameworkDefaults.props" />
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.proj:23]: I don't see this import being used anywhere below. I don't think that the Arcade TFM properties can be used.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.proj [hunk 6]
```diff
@@ -57,6 +58,9 @@
 
       <!-- SN is only available on Windows -->
       <SNBinaryPath Condition="$([MSBuild]::IsOSPlatform('Windows'))">$(NuGetPackageRoot)sn\$(SNVersion)\sn.exe</SNBinaryPath>
+      
+      <!-- .pkgs and .app bundle tooling is only available on MacOS -->
+      <PkgToolPath Condition="$([MSBuild]::IsOSPlatform('OSX'))">$(NuGetPackageRoot)microsoft.dotnet.macospkg\$(MicrosoftDotNetMacOsPkgVersion)\tools\net10.0\any\Microsoft.Dotnet.MacOsPkg.dll</PkgToolPath>
```
Comment (ellahathaway) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.proj:63]: Is there a way to use a prop for net10.0? I think the hardcoded version will cause issues when we eventually update the target framework.

File: src/Microsoft.DotNet.MacOsPkg/AppBundle.cs [hunk 7]
```diff
@@ -1,10 +1,49 @@
...
+                        // If we hit this, and it is a valid app bundle, then we should adjust the logic above.
+                        // If we never hit this (likely), then remove.
+                        throw new Exception("Unexpected .app directory structure. Please contact dnceng.");
+                    }
+                    return containsAppBundleContents;
+                }
+            }
+            else
+            {
+                return Path.GetExtension(path) == ".app";
```
Comment (ellahathaway) [inline src/Microsoft.DotNet.MacOsPkg/AppBundle.cs:41]: Is it ever the case where we are repacking a .pkg and encounter something like "microsoft.netcore.app" (not a zipped .app directory)? If so, maybe we should check the file type (a zip) in addition to the extension?

File: src/Microsoft.DotNet.SignTool/src/BatchSignUtil.cs [hunk 8]
```diff
@@ -126,17 +124,20 @@ private bool SignFiles()
             bool signGroup(IEnumerable<FileSignInfo> files, out int signedCount)
             {
                 var filesToSign = files.Where(fileInfo => fileInfo.SignInfo.ShouldSign).ToArray();
+                var filesToNotarize = files.Where(fileInfo => fileInfo.SignInfo.ShouldNotarize).ToArray();
                 signedCount = filesToSign.Length;
                 if (filesToSign.Length == 0) return true;
```
Comment (ellahathaway) [inline src/Microsoft.DotNet.SignTool/src/BatchSignUtil.cs:129]: Are we ever notarizing but not signing?

File: src/Microsoft.DotNet.SignTool/src/BatchSignUtil.cs [hunk 9]
```diff
@@ -548,84 +568,96 @@ private void VerifyCertificates(TaskLoggingHelper log)
...
-            }
-            else if (file.IsPowerShellScript())
-            {
-                if (!_signTool.VerifySignedPowerShellFile(file.FullPath))
+                else if (file.IsPkg() || file.IsAppBundle())
                 {
-                    _log.LogError($"Powershell file {file.FullPath} does not have a signature mark.");
+                    if (!_signTool.VerifySignedPkgOrAppBundle(file.FullPath, _signTool.PkgToolPath))
+                    {
+                        _log.LogError($"Powershell file {file.FullPath} does not have a signature marker.");
```
Comment (ellahathaway) [inline src/Microsoft.DotNet.SignTool/src/BatchSignUtil.cs:647]:
```suggestion
                        _log.LogError($"Pkg or app file {file.FullPath} does not have a signature marker.");
```

File: src/Microsoft.DotNet.SignTool/src/BatchSignUtil.cs [hunk 10]
```diff
@@ -548,84 +568,96 @@ private void VerifyCertificates(TaskLoggingHelper log)
...
+                    }
+                    else
+                    {
+                        _log.LogMessage(MessageImportance.Low, $"Powershell file {file.FullPath} has a signature marker.");
```
Comment (ellahathaway) [inline src/Microsoft.DotNet.SignTool/src/BatchSignUtil.cs:651]:
```suggestion
                        _log.LogMessage(MessageImportance.Low, $"Pkg or app file {file.FullPath} has a signature marker.");
```

File: src/Microsoft.DotNet.SignTool/src/ZipData.cs [hunk 11]
```diff
@@ -289,6 +302,97 @@ private void RepackWixPack(TaskLoggingHelper log, string tempDir, string wixTool
             }
         }
 
+        internal static bool RunPkgProcess(string srcPath, string dstPath, string action, string pkgToolPath)
+        {
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            {
+                throw new Exception($"Pkg tooling is only support on MacOS.");
```
Comment (ellahathaway) [inline src/Microsoft.DotNet.SignTool/src/ZipData.cs:309]:
```suggestion
                throw new Exception($"Pkg tooling is only supported on MacOS.");
```

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.props [hunk 12]
```diff
@@ -60,6 +63,13 @@
     <FileExtensionSignInfo Include=".zip" CertificateName="None" />
     <FileExtensionSignInfo Include=".tgz" CertificateName="None" />
     <FileExtensionSignInfo Include=".tar.gz" CertificateName="None" />
+    <!-- Note, these can only be unpack/repacked on Mac. They can be signed on a non-Mac -->
```
Comment (ellahathaway) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.props:66]: We can sign pkgs on non-mac, but we cannot verify that they were signed.

As is, the signing verification will throw an exception on non-mac because the verification uses the pkgtool. We should either adjust the code to not allow pkgs to be signed on non-mac, or we should not throw when validating pkg signatures on non-mac.

File: src/Microsoft.DotNet.SignTool/src/BatchSignUtil.cs (hunk 8 above)
Comment (mmitche) [reply src/Microsoft.DotNet.SignTool/src/BatchSignUtil.cs:129]: I think it's highly unlikely, and probably not supported right now. We do regularly sign without notarizing though. I wanted to indicate that there were two operations going on.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.proj (hunk 5 above)
Comment (mmitche) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.proj:23]: I think I had this in there to use NetToolCurrent for the command line tools. Updated

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.proj (hunk 6 above)
Comment (mmitche) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.proj:63]: $(NetToolCurrent). I'll try it.

File: src/Microsoft.DotNet.MacOsPkg/AppBundle.cs (hunk 7 above)
Comment (mmitche) [reply src/Microsoft.DotNet.MacOsPkg/AppBundle.cs:41]: I think I see what you're getting at?

I don't think there are any files we would encounter which would be non-dir .app files that are also not zips. We'll implicitly get this check because during repacking we would end up trying to unzip the .app file. If it wasn't a zip, it would fail.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.props (hunk 12 above)
Comment (mmitche) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.props:66]: Good catch. I think it's worth keeping the signing, since someone could be set up to sign on Windows. I'll no-op the verification.

File: src/Microsoft.DotNet.SignTool/ItemToSign.cs
Comment (NikolaMilosavljevic) [inline src/Microsoft.DotNet.SignTool/ItemToSign.cs:1]: Is there a reason for this file to be in project root and not in `src`?

Comment (mmitche) [reply src/Microsoft.DotNet.SignTool/ItemToSign.cs:1]: Because VS put it there. Thanks!

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.props (hunk 12 above)
Comment (mmitche) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.props:66]: After looking a bit more, it's good to note that we can sign pkgs or .dylibs or whatever, but we can't unpack/repack pkgs or verify any signing. So we have two options:
- Sign the files on non-Mac without recursive unpacking/packing. This has the potential to be confusing to the user, but is less breaking compared to current behavior.
- Error and tell people to move to Mac

Comment (ellahathaway) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.props:66]: I lean towards the first option. I think if we error, we risk breaking product repos that currently sign pkgs.

Comment (ellahathaway) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.props:66]: The one thing to be caution of here is "double signing."

To my knowledge, while verifying the signature of these files, we need to return "false" on non-mac. Otherwise, the file will be considered "signed" and will not be submitted for signing, regardless of if it's currently signed or not.

The issue here is that a file may be resubmitted, and I do not know if that will cause errors/failures.

Comment (mmitche) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.props:66]: Good point. I'll add a couple windows-only tests for this. Generally signatures can be replaced. Some signatures are specified as being 'dual sign', where they are only usable as an additional signature (3rdpartySha2)

**Review by ellahathaway — COMMENTED** [review]
I cannot approve because I’m technically the author of the PR, but the changes look good to me. Thanks for all the work you did on this, Matt.

File: src/Microsoft.DotNet.SignTool/src/Configuration.cs [hunk 13]
```diff
@@ -569,6 +541,39 @@ private FileSignInfo ExtractSignInfo(
             }
 
             return new FileSignInfo(file, SignInfo.Ignore, wixContentFilePath: wixContentFilePath);
+
+            bool IsSignedWithLogging(PathWithHash file, SigningStatus signingStatus)
```
Comment (NikolaMilosavljevic) [inline src/Microsoft.DotNet.SignTool/src/Configuration.cs:545]: Perhaps `WithLogging` suffix isn't necessary? Same for `IsStrongNameSignedWithLogging`

**Comment by mmitche** [comment]
**Test build with mac signing enabled:** https://dev.azure.com/dnceng/internal/_build/results?buildId=2622013&view=logs&j=e0ba34c9-23d1-568a-f8ed-6319140a5963&t=e0ba34c9-23d1-568a-f8ed-6319140a5963
**aspnetcore build with new sign tool:** https://dev.azure.com/dnceng/internal/_build/results?buildId=2622014&view=results
**sdk build with new sign tool:** https://dev.azure.com/dnceng/internal/_build/results?buildId=2622037&view=results

**Review by NikolaMilosavljevic — APPROVED** [review]
